tests/components/tasmota/test_fan.py

```diff
@@ -0,0 +1,255 @@
+    help_test_entity_id_update_subscriptions,
+)
+
+from tests.async_mock import patch
+from tests.common import async_fire_mqtt_message
+from tests.components.fan import common
+
+
+async def test_controlling_state_via_mqtt(hass, mqtt_mock, setup_tasmota):
+    """Test state update via MQTT."""
+    config = copy.deepcopy(DEFAULT_CONFIG)
+    config["if"] = 1
+    mac = config["mac"]
+
+    async_fire_mqtt_message(
+        hass,
+        f"{DEFAULT_PREFIX}/{mac}/config",
+        json.dumps(config),
+    )
+    await hass.async_block_till_done()
+
+    state = hass.states.get("fan.tasmota")
+    assert state.state == "unavailable"
+    assert not state.attributes.get(ATTR_ASSUMED_STATE)
+
+    async_fire_mqtt_message(hass, "tasmota_49A3BC/tele/LWT", "Online")
+    state = hass.states.get("fan.tasmota")
+    assert state.state == STATE_OFF
+    assert state.attributes["speed"] is None
+    assert state.attributes["speed_list"] == ["off", "low", "medium", "high"]
+    assert state.attributes["supported_features"] == fan.SUPPORT_SET_SPEED
+    assert not state.attributes.get(ATTR_ASSUMED_STATE)
+
+    async_fire_mqtt_message(hass, "tasmota_49A3BC/tele/STATE", '{"FanSpeed":1}')
+    state = hass.states.get("fan.tasmota")
+    assert state.state == STATE_ON
+    assert state.attributes["speed"] == "low"
+
+    async_fire_mqtt_message(hass, "tasmota_49A3BC/tele/STATE", '{"FanSpeed":2}')
+    state = hass.states.get("fan.tasmota")
+    assert state.state == STATE_ON
+    assert state.attributes["speed"] == "medium"
+
+    async_fire_mqtt_message(hass, "tasmota_49A3BC/tele/STATE", '{"FanSpeed":3}')
+    state = hass.states.get("fan.tasmota")
+    assert state.state == STATE_ON
+    assert state.attributes["speed"] == "high"
+
+    async_fire_mqtt_message(hass, "tasmota_49A3BC/tele/STATE", '{"FanSpeed":0}')
+    state = hass.states.get("fan.tasmota")
+    assert state.state == STATE_OFF
+    assert state.attributes["speed"] == "off"
+
+    async_fire_mqtt_message(hass, "tasmota_49A3BC/stat/RESULT", '{"FanSpeed":1}')
+    state = hass.states.get("fan.tasmota")
+    assert state.state == STATE_ON
+    assert state.attributes["speed"] == "low"
+
+    async_fire_mqtt_message(hass, "tasmota_49A3BC/stat/RESULT", '{"FanSpeed":0}')
+    state = hass.states.get("fan.tasmota")
+    assert state.state == STATE_OFF
+    assert state.attributes["speed"] == "off"
+
+
+async def test_sending_mqtt_commands(hass, mqtt_mock, setup_tasmota):
+    """Test the sending MQTT commands."""
+    config = copy.deepcopy(DEFAULT_CONFIG)
+    config["if"] = 1
+    mac = config["mac"]
+
+    async_fire_mqtt_message(
+        hass,
+        f"{DEFAULT_PREFIX}/{mac}/config",
+        json.dumps(config),
+    )
+    await hass.async_block_till_done()
+
+    async_fire_mqtt_message(hass, "tasmota_49A3BC/tele/LWT", "Online")
+    state = hass.states.get("fan.tasmota")
+    assert state.state == STATE_OFF
+    await hass.async_block_till_done()
+    await hass.async_block_till_done()
```

Comment:
Are both of these needed? (e.g. callbacks invoking more callbacks?)

Reply:
Yeah, that's right.